src/history/hs.c

```diff
@@ -763,13 +773,16 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_PAGE *page, WT_MULTI *multi)
                  * It is not correct to check prev_upd == list->onpage_upd as we may have aborted
                  * updates in the middle.
                  *
-                 * We can't calculate reverse modify based on an uncommitted prepared update because
-                 * it may be aborted.
+                 * We must insert the first and second updates after a prepared update as full
+                 * values because if the prepared update is aborted, we will remove the first update
+                 * after it from the history store to the update chain. Readers reading the older
+                 * values need a full update as the base value for constructing reverse modifies.
```

Comment:
I think we need to check this. Even we will retry later anyway, we may seg fault because we apply modify on the wrong base value.

Reply:
I think we have the code to refer the on-disk value, why we need this change?

Reply:
Suppose we have a prepare on the disk, then one full update and another modify in the history store. Reader decides to read the modify. While the same time the full update in the history store is removed. We will base the modify on the onpage value, which is wrong. Though anyway we may retry, but when we applying the modify, there is a risk we may seg fault.

Reply:
Perhaps we can check for the race before we reconstruct the modify. When we're looking and doing `__wt_modify_apply_item`, we're not reading from history store anymore. So we should be able to check whether we raced and if not, reconstruct.

Reply:
Yeah, that sounds better.

Reply:
mhh, no the race can happen when we are reconstructing.

Reply:
The downside of that is the logic become more scattered. I'm not sure on that.

Reply:
Sure, I get your point. It's not 100% clear cut to me either.

Although I think that reconstructing the second update behind a prepare is pretty gnarly and not free of technical debt either. And it means that modifies will be less effective in the presence of prepared updates.

Reply:
@tetsuo-cpp I think your idea is not working because once the history store value is deleted. The search is not going to be successful before reconciliation runs as the base value of that modify is now on the update chain only.

Reply:
@quchenhao 
Yeah, I understand. It'd have to consider the update chain also for that to work which would be a big change.